js/ui/windowManager.js

```diff
@@ -2166,3 +1888,368 @@ const WindowManager = new Lang.Class({
         }
     },
 });
+
+const ICON_BOUNCE_MAX_SCALE = 0.4;
+const ICON_BOUNCE_ANIMATION_TIME = 1.0;
+const ICON_BOUNCE_ANIMATION_TYPE_1 = 'easeOutSine';
+const ICON_BOUNCE_ANIMATION_TYPE_2 = 'easeOutBounce';
+
+const BUTTON_OFFSET_X = 100;
+const BUTTON_OFFSET_Y = 100;
+
+const SessionState = {
+    APP: 0,
+    BUILDER: 1
+};
+
+function animateBounce(actor)
+{
+    Tweener.removeTweens(actor);
+    if (!Tweener.isTweening(actor)) {
+        Tweener.addTween(actor, {
+                scale_y: 1 + ICON_BOUNCE_MAX_SCALE,
+                scale_x: 1 + ICON_BOUNCE_MAX_SCALE,
+                translation_y: actor.height * ICON_BOUNCE_MAX_SCALE,
+                translation_x: actor.width * ICON_BOUNCE_MAX_SCALE / 2,
+                time: ICON_BOUNCE_ANIMATION_TIME * 0.25,
+                delay: 0.3,
+                transition: ICON_BOUNCE_ANIMATION_TYPE_1
+        });
+        Tweener.addTween(actor, {
+                scale_y: 1,
+                scale_x: 1,
+                translation_y: 0,
+                translation_x: 0,
+                time: ICON_BOUNCE_ANIMATION_TIME * 0.75,
+                transition: ICON_BOUNCE_ANIMATION_TYPE_2,
+                delay: ICON_BOUNCE_ANIMATION_TIME * 0.25 + 0.3,
+                onComplete: function() {animateBounce(actor);}
+        });
+    }
+}
+
+const CodingManager = new Lang.Class({
+    Name: 'CodingManager',
+
+    _init: function(actor) {
+        this._sessions = [];
+        this._rotateInActors = [];
+        this._rotateOutActors = [];
+        this._firstFrameConnections = [];
+    },
+
+    addSwitcherToBuilder: function(actorApp) {
+        let window = actorApp.get_meta_window();
+
+        let button = new St.Button({ style_class: 'view-source' });
+        let rect = window.get_frame_rect();
+        button.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        Main.layoutManager.addChrome(button);
+
+        this._sessions.push({buttonApp: button,
+                             windowApp: actorApp,
+                             state: SessionState.APP});
+
+        let session = this._getSession(actorApp);
+        button.connect('clicked', Lang.bind(this, this._switchToBuilder, session));
+        let positionChangedId = window.connect('position-changed', Lang.bind(this, this._windowAppPositionChanged, session));
+        let windowsRestackedId = Main.overview.connect('windows-restacked', Lang.bind(this, this._windowAppRestacked, session));
+        session.positionChangedIdApp = positionChangedId;
+        session.windowsRestackedIdApp = windowsRestackedId;
+    },
+
+    addSwitcherToApp: function(actorBuilder, windowApp) {
+        let currentSession = this._sessions.filter(function(session) {
+            return (session.windowApp.get_meta_window() === windowApp);
+        });
+        if (currentSession.length === 0)
+            return;
+
+        let session = currentSession[0];
```

Comment:
Could have a `getCodingSessionForWindowApp` helper here to avoid repeating this.